render-ws-java-client/src/main/java/org/janelia/render/client/MultiSemFlatFieldCorrectionClient.java

```diff
@@ -0,0 +1,195 @@
+		@Parameter(names = "--targetStackSuffix", description = "Suffix to append to the stack name for the corrected stack", required = true)
+		private String targetStackSuffix = "_corrected";
+		@Parameter(names = "--inputRoot", description = "Root folder for input data; if given, the structure under this root is replicated in the output folder")
+		private String inputRoot = null;
+		@Parameter(names = "--flatFieldConstantFromZ", description = "Maximum z-layer of flat field estimates to consider. All subsequent z-layers get corrected with the maxium, since the estimates can be bad for the last few z-layers. If not given, all z-layers are considered")
+		private Integer flatFieldConstantFromZ = Integer.MAX_VALUE;
+	}
+
+	public static void main(final String[] args) {
+		final ClientRunner clientRunner = new ClientRunner(args) {
+			@Override
+			public void runClient(final String[] args) throws IOException {
+
+				final Parameters parameters = new Parameters();
+				parameters.parse(args);
+				LOG.info("runClient: entry, parameters={}", parameters);
+
+				final MultiSemFlatFieldCorrectionClient client = new MultiSemFlatFieldCorrectionClient(parameters);
+				client.correctTiles();
+			}
+		};
+		clientRunner.run();
+	}
+
+	public MultiSemFlatFieldCorrectionClient(final Parameters parameters) {
+		this.params = parameters;
+		this.renderClient = new RenderDataClient(parameters.multiProject.baseDataUrl, parameters.multiProject.owner, parameters.multiProject.project);
+	}
+
+	public void correctTiles() throws IOException {
+		final List<StackId> stacks = params.multiProject.stackIdWithZ.getStackIdList(renderClient);
+
+		// TODO: iterate by z-layer first to re-use the flat field estimate for all stacks?
+		for (final StackId stack : stacks) {
+			final DoubleSummaryStatistics zStatistics = getZStatistics(stack);
+			final StackMetaData stackMetaData = renderClient.getStackMetaData(stack.getStack());
+			renderClient.setupDerivedStack(stackMetaData, stack.getStack() + params.targetStackSuffix);
+
+			for (int z = (int) zStatistics.getMin(); z <= (int) zStatistics.getMax(); z++) {
+				final ResolvedTileSpecCollection tileSpecs = renderClient.getResolvedTiles(stack.getStack(), (double) z);
+
+				for (final TileSpec tileSpec : tileSpecs.getTileSpecs()) {
+					final ImageProcessor ip = loadImageTile(tileSpec);
+					final int sfov = extractSfovNumber(tileSpec);
+					final ImageProcessor flatFieldEstimate = loadFlatFieldEstimate(Math.min(z, params.flatFieldConstantFromZ), sfov);
+
+					applyFlatFieldCorrection(ip, flatFieldEstimate);
+
+					patchTileSpec(tileSpec);
+					saveImage(ip, tileSpec);
+				}
+
+				renderClient.saveResolvedTiles(tileSpecs, stack.getStack() + params.targetStackSuffix, (double) z);
+			}
+			renderClient.setStackState(stack.getStack() + params.targetStackSuffix, StackMetaData.StackState.COMPLETE);
+		}
+	}
+
+	private DoubleSummaryStatistics getZStatistics(final StackId stack) throws IOException {
```

Comment:
I didn't see the need for this

Reply:
That is a remainder from when I wanted to compute the min and max for all the stacks up front. You are right, this is not needed anymore and I'll refactor this to just use the stack's own z-values.